collector/ipvs.go

```diff
@@ -0,0 +1,321 @@
+			Name:      "incoming_bytes_total",
+			Help:      "The total amount of incoming data.",
+		},
+	)
+	c.outgoingBytes = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: Namespace,
+			Subsystem: ipvsSubsystem,
+			Name:      "outgoing_bytes_total",
+			Help:      "The total amount of outgoing data.",
+		},
+	)
+
+	c.backendConnectionsActive = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: Namespace,
+			Subsystem: ipvsSubsystem,
+			Name:      "backend_connections_active",
+			Help:      "The current active connections by local and remote address.",
+		},
+		ipvsBackendLabelNames,
+	)
+	c.backendConnectionsInact = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: Namespace,
+			Subsystem: ipvsSubsystem,
+			Name:      "backend_connections_inactive",
+			Help:      "The current inactive connections by local and remote address.",
+		},
+		ipvsBackendLabelNames,
+	)
+	c.backendWeight = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: Namespace,
+			Subsystem: ipvsSubsystem,
+			Name:      "backend_weight",
+			Help:      "The current backend weight by local and remote address.",
+		},
+		ipvsBackendLabelNames,
+	)
+
+	return &c, nil
+}
+
+func (c *ipvsCollector) Update(ch chan<- prometheus.Metric) error {
+	ipvsStats, err := getIPVSStats()
+	if err != nil {
+		return fmt.Errorf("could not get IPVS stats: %s", err)
+	}
+
+	c.connections.Set(float64(ipvsStats.connections))
+	c.incomingPackets.Set(float64(ipvsStats.incomingPackets))
+	c.outgoingPackets.Set(float64(ipvsStats.outgoingPackets))
+	c.incomingBytes.Set(float64(ipvsStats.incomingBytes))
+	c.outgoingBytes.Set(float64(ipvsStats.outgoingBytes))
+
+	c.connections.Collect(ch)
+	c.incomingPackets.Collect(ch)
+	c.outgoingPackets.Collect(ch)
+	c.incomingBytes.Collect(ch)
+	c.outgoingBytes.Collect(ch)
+
+	backendStats, err := getIPVSBackendStatus()
+	if err != nil {
+		return fmt.Errorf("could not get backend status: %s", err)
+	}
+
+	for _, backend := range backendStats {
+		labelValues := []string{
+			backend.localAddress.String(),
+			strconv.FormatUint(uint64(backend.localPort), 10),
+			backend.remoteAddress.String(),
+			strconv.FormatUint(uint64(backend.remotePort), 10),
+			backend.proto,
+		}
+		c.backendConnectionsActive.WithLabelValues(labelValues...).Set(float64(backend.activeConn))
+		c.backendConnectionsInact.WithLabelValues(labelValues...).Set(float64(backend.inactConn))
+		c.backendWeight.WithLabelValues(labelValues...).Set(float64(backend.weight))
+	}
+
+	c.backendConnectionsActive.Collect(ch)
+	c.backendConnectionsInact.Collect(ch)
+	c.backendWeight.Collect(ch)
+
+	return nil
+}
+
+func getIPVSBackendStatus() ([]ipvsBackendStatus, error) {
+	file, err := os.Open(procNetIPVS)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return parseIPVSBackendStatus(file)
+}
+
+func parseIPVSBackendStatus(file io.Reader) ([]ipvsBackendStatus, error) {
+	var (
+		status       = []ipvsBackendStatus{}
```

Comment:
Just one more hint for the future that I didn't notice on the first pass, and which isn't terribly important:

I'd write this either as:

`var status []ipvsBackendStatus` (no assignment and construction needed in this case - the zero value is fine)

...or as:

`status := []ipvsBackendStatus{}` (without any `var`)